Feat/create event

GagahGK opened the pull request:
a new endpoint to create webinar / inlive event

changelogs : 
- Added new component for event related interaction 
   - Date Picker using [react-datepicker](https://reactdatepicker.com/)
   - Time Picker
   - Image Cropper using [react-cropper](https://github.com/react-cropper/react-cropper)
- new page to create event 
   - require sign in to access the page
   - uses [compressor-js ](https://www.npmjs.com/package/compressorjs) before uploading the image to server
   - added new column to the event table 
      - `updated_at` 
      - `is_published` 
      - `thumbnail_url`
   - only event owner can access the event if event is not yet published
- uses custom server with express to host the event thumbnail
- updated next config & dockerfile to build app using custom server
    - next doesn't perform build with standalone

faiq-naufal commented:
@GagahGK For some reason, I cannot get env `INLIVE_HUB_API_KEY` value after building and running the custom server through docker. This caused the website cannot create a new room when running through docker. Can you check this?

faiq-naufal commented on .env.local.example:
```diff
@@ -36,3 +36,5 @@ ENABLE_MAILER=false
 MAILER_DOMAIN=
 MAILER_API_KEY=
 ROOM_INV_EMAIL_TEMPLATE=
+
+ROOM_PERSISTANT_VOLUME_PATH=./volume
```
Don't forget to add other envs in this example. For example, `PORT` and `LOCAL_STORAGE_PATH` haven't been specified here

faiq-naufal commented on next.config.js:
```diff
@@ -2,7 +2,7 @@ const { withSentryConfig } = require('@sentry/nextjs');
 /** @type {import('next').NextConfig} */
 const nextConfig = {
-  output: 'standalone',
+  // output: 'standalone',
```
Can you tell me the reason to turn off the `standalone` output when using the custom server?

faiq-naufal commented on app/_features/event/styles/bg.png:
If the bg is an icon. Better to use svg format. Also, if this icon is not reusable, better to put and embed it besides other html code as svg.

GagahGK replied:
this one is necessary file that is used only for the [react-cropper](https://github.com/react-cropper/react-cropper) library, i'm not sure if we can change the change it to svg without changing the react-cropper library code

faiq-naufal replied:
I see, then I think it's fine to leave it as it is for now

faiq-naufal commented on package.json:
```diff
@@ -47,10 +57,12 @@
     "eslint-config-prettier": "^8.8.0",
     "eslint-plugin-prettier": "^4.2.1",
     "eslint-plugin-tailwindcss": "^3.13.0",
+    "nodemon": "^3.0.3",
```
Will we use nodemon for custom server? I don't see any nodemon command in the running script.

GagahGK replied:
we actually don't need this, i'm going to remove it

GagahGK commented:
> @GagahGK For some reason, I cannot get env `INLIVE_HUB_API_KEY` value after building and running the custom server through docker. This caused the website couldn't create a new room when running through docker in my side. Can you check this? This doesn't happen in the main branch version. Probably because the main branch is using standalone output.

I think can solve this by building the code using the standalone option, previously there was a warning and I'm not sure how to handle it, turns out I just need to replace the provided server.js file with the custom server.js

but still the node_modules need to be copied since the standalone didn't provide express library

GagahGK commented:
@faiq-naufal sorry for taking a while to confirm case, but I've tried to use both standalone and non standalone build through docker using docker-compose the server able to create the meeting room successfully

have you make sure your API key is not expired ?

faiq-naufal commented:
@GagahGK  Ah sorry, I'm not talking about running with docker compose. I'm talking about building the image and running the image manually like in this steps. https://github.com/inlivedev/inlive-room#running-with-docker. The environment variables when using custom server are not included if I run using those steps. In the `main` branch, I still get the environment variables when running with those steps.

GagahGK commented:
@faiq-naufal 

it looks like that the default server.js uses the generated .env and loaded it during runtime, but I think it won't be a problem since the image will be loaded into a kubernetes pod. we can update the config file in kubernetes to load the env variable.

but due to this to run in local now we need to pass a environment argument to load a variable 

```
docker run -p 3000:3000 -e VARIABLE=value inlive-room:latest
```

faiq-naufal commented:
@GagahGK I see then, I think it's fine if this is not an issue with the deployment.

faiq-naufal commented on package.json:
```diff
@@ -2,11 +2,15 @@
   "name": "room.inlive.app",
   "version": "0.1.0",
   "private": true,
+  "type": "commonjs",
```
We don't need to specify the type actually. By default, when type is not set, it is treated as commonjs. See https://nodejs.org/api/packages.html#packagejson-and-file-extensions

faiq-naufal commented on package.json:
```diff
@@ -51,6 +61,7 @@
     "tailwindcss": "^3.4.1",
-    "typescript": "^5.3.3"
+    "ts-node": "^10.9.2",
+    "typescript": "^5.1.6"
```
Why did you downgrade the typescript version to `5.1.6`?

faiq-naufal commented on package.json:
```diff
@@ -2,11 +2,15 @@
-    "dev": "next dev",
-    "build": "next build",
-    "start": "next start",
+    "dev:custom-server": "ts-node --project tsconfig.server.json ./app/server.ts",
+    "dev:next": "next dev",
+    "build:server": "tsc --project tsconfig.server.json",
+    "build:next": "next build",
+    "build": "npm run build:next && npm run build:server",
+    "start": "NODE_ENV=production node dist/server.js",
```
Do we actually still need `dev:next` script? I think we don't need this script.
I think we can make the `dev` and `build` scripts simpler. Below scripts are much simpler. We don't need to update the readme too.

```
{
    "dev": "ts-node --project tsconfig.server.json ./app/server.ts",
    "build": "next build && tsc --project tsconfig.server.json",
}
```

faiq-naufal replied:
@GagahGK `dev:custom-server` should be changed to `dev` as default development script. Then, whenever we want to use built-in server, we can use the `dev:next`. For the `build` script, there is no need to separate it as `build:server` and `build:next` scripts

GagahGK replied:
then if that's the case then it's agreeable, but we might need to add a note about `dev` uses a custom server and its limitation about the hot reload ability

faiq-naufal replied:
Sure, better to give a little info about the custom server

faiq-naufal commented on app/_shared/components/icons/chevron-right.tsx:
```diff
@@ -0,0 +1,18 @@
+export default function ChevronRight(props: SVGElementPropsType) {
+  return (
+    <svg
+      width="24"
+      height="24"
```
Again, this makes icon size is hard-coded to 24x24 size. Use {...props} instead.

faiq-naufal commented on app/_features/event/styles/cropper.css:
@GagahGK  Why did you not import `cropperjs/dist/cropper.css` from node_modules instead?

faiq-naufal commented on app/_features/event/styles/image-cropper.css:
@GagahGK  This file is not needed isn't it? There is no file which imports this file

faiq-naufal commented on app/_features/event/components/image-cropper.tsx:
```diff
@@ -0,0 +1,128 @@
+          type: 'ConfirmCrop',
+          payload: { blob: blob, preview: URL.createObjectURL(blob) },
+        });
+      } else console.log('blob is null');
```
Remove this else console.log()

GagahGK replied:
I've removed the logging https://github.com/inlivedev/inlive-room/pull/130/commits/06de4972dc0b8b36258652c019bd4a9acd9245ff#diff-d0ca42e2c8f5450a47f15205e8d83111bd45cee06f0be0ff3915367d6d4c6988L78-L80

faiq-naufal commented on app/_features/event/components/image-cropper.tsx:
```diff
@@ -0,0 +1,128 @@
+
+  const getCropData = useCallback(async () => {
+    if (typeof cropperRef.current?.cropper !== 'undefined') {
+      if (!cropperRef.current) console.log('cropperRef.current is undefined');
```
No need to check and console.log

GagahGK replied:
removed the checking https://github.com/inlivedev/inlive-room/pull/130/commits/06de4972dc0b8b36258652c019bd4a9acd9245ff#diff-d0ca42e2c8f5450a47f15205e8d83111bd45cee06f0be0ff3915367d6d4c6988R35-R37

faiq-naufal commented on app/_features/event/components/event-date-picker.tsx:
```diff
@@ -0,0 +1,84 @@
+            onChange={(date) => {
+              if (date) {
+                setSelectedDate(date);
+                console.log(setSelectedDate);
```
No need to log the selected date

GagahGK replied:
i've removed the logging https://github.com/inlivedev/inlive-room/pull/130/commits/06de4972dc0b8b36258652c019bd4a9acd9245ff#diff-e67452b9ca88fae0ce442b172f51c8f3721a58438e933fd32db8bcd31006655aL62

faiq-naufal commented on app/_features/event/components/event-form.tsx:
```diff
@@ -0,0 +1,493 @@
+          window.location.origin
+        ).href;
+        navigateTo(redirectPath);
+      } else {
+        console.log(respEvent.message);
+      }
```
No need to log the message with else statement. Use alert instead if you want to notify the user.

GagahGK replied:
removed the console log and replaced it with an alert instead 
removed the unnecessary console log https://github.com/inlivedev/inlive-room/pull/130/commits/06de4972dc0b8b36258652c019bd4a9acd9245ff#diff-9e2bd548722a266fc9bcce2be48f3d5b8b48215f5adbedbfbffa138083b1e33eL193-R184

faiq-naufal commented on app/_features/event/components/event-form.tsx:
```diff
@@ -0,0 +1,493 @@
+  const { isOpen, onOpen, onClose, onOpenChange } = useDisclosure();
+
+  const openModal = useCallback(() => {
+    console.log('open coming soon modal');
```
Remove this logging

GagahGK replied:
removed the unnecessary console log https://github.com/inlivedev/inlive-room/pull/130/commits/06de4972dc0b8b36258652c019bd4a9acd9245ff

faiq-naufal commented on app/_features/event/components/event-form.tsx:
```diff
@@ -0,0 +1,493 @@
+    createEvent(true);
+  }, [createEvent]);
+
+  const updateDraft = useCallback(() => {
+    console.log('update draft');
+  }, []);
+
+  const onPublish = useCallback(() => createEvent(true), [createEvent]);
+
+  const updatePublishedEvent = useCallback(() => {
+    console.log('update published event');
+  }, []);
```
If you intend to  continue `updateDraft` and `updatePublishedEvent` later, I think better to change this with descriptive comment.

GagahGK replied:
updated with comments instead : https://github.com/inlivedev/inlive-room/pull/130/commits/ecb61927672c4fa4bab051eaba5a36f3f471bc4d

faiq-naufal reviewed (changes requested):
Some logging need to be removed

faiq-naufal commented on app/_features/event/components/event-form.tsx:
```diff
@@ -0,0 +1,493 @@
+    []
+  );
+
+  useEffect(() => {
+    if (!user) document.dispatchEvent(new CustomEvent('open:sign-in-modal'));
+  }, [user]);
```
You shouldn't check if the user has logged in or not in this component. This component should be only rendered/displayed when the user actually exists. This logic should be outside this component. For example, when the user is not logged in, the user should not be able to access the create event or update event pages.

GagahGK replied:
I've limited the access to the page by showing a http error instead https://github.com/inlivedev/inlive-room/pull/130/commits/7c9946511cbc67deaeae121f6a1f4d4d4935bc88

faiq-naufal commented on app/_features/event/components/event-form.tsx:
```diff
@@ -0,0 +1,493 @@
+  }
+};
+
+const APP_ORIGIN = process.env.NEXT_PUBLIC_APP_ORIGIN;
+const LOCAL_STORAGE_PATH = process.env.LOCAL_STORAGE_PATH;
```
These two are unused variables.

faiq-naufal commented:
@GagahGK I simplified the server.ts logic on [76863eb](https://github.com/inlivedev/inlive-room/pull/130/commits/76863eb3c39225a4bce232875e5532077a546ad6). No need to throw an error when envs are not set and give them default value instead. I also added env to the deployment workflow [fefc272](https://github.com/inlivedev/inlive-room/pull/130/commits/fefc2726b82379a73b4957445ccf15f6a37a69ec).

GagahGK commented:
@faiq-naufal thankyou, I've updated the code with your requested changes : 
- removing the console logging
- limit access to the create event page
- added comments to unimplemented `updateDraft()` function
- reverting the typescript version and
- making the icon scalable

faiq-naufal commented:
@GagahGK currently, there is no navigation link to the event create page `/event/create`. I think this is fine for now considering there are more features such as event list, update event that haven't been developed. But, I want to hear your opinion about this. It seems, mas Ashar also hasn't built the UI for the navigation link to the event create page.

GagahGK commented:
> @GagahGK currently, there is no navigation link to the event create page `/event/create`. I think not having navigation link is fine for now considering there are more features such as event list, update event that haven't been finished. But, I want to hear your thought about this. It seems, mas Ashar also hasn't built the UI for the navigation link to the event create page.

@faiq-naufal I also noticed that and have asked him previously and haven't receive any update, I think we should do a follow up.
but I think, it was mean to be navigated through the `My Events` page as seen on the figma

faiq-naufal commented:
@GagahGK Ah yeah, that's correct. We need to finish the event list page first and from there user can create a new event.

GagahGK commented:
> @GagahGK Ah yeah, that's correct. We need to finish the "My Events" page first and from there user can create a new event.

@faiq-naufal it'll be delivered with the next branch alongside with the update & delete event feature
